lib/ansible/modules/cloud/xenserver/xenserver_guest.py

```diff
@@ -0,0 +1,1510 @@
+
+        changed_state = False
+
+        try:
+            # First check the current state of the VM.
+            vm_power_state = self.vm_params['power_state'].lower()
+
+            # Support for ansible check mode.
+            if vm_power_state in ["paused", "running"] and self.module.check_mode:
+                return True
+
+            # hard_restart will restart VM regardless of current state.
+            if vm_power_state in ["paused", "running"]:
+                self.xapi_session.xenapi.VM.hard_reboot(self.vm_ref)
+                self.vm_params['power_state'] = "running"
+                changed_state = True
+            else:
+                self.module.fail_json(msg="Cannot restart VM in state '%s'!" % vm_power_state)
+        except XenAPI.Failure as f:
+            self.module.fail_json(msg="XAPI ERROR: %s" % f.details)
+
+        return changed_state
+
+    def shutdownguest(self):
+        # Safety check.
+        if self.vm_ref is None:
+            self.module.fail_json(msg="Called shutdownguest on non existing VM!")
+
+        changed_state = False
+
+        try:
+            # First check the current state of the VM.
+            vm_power_state = self.vm_params['power_state'].lower()
+
+            # Support for ansible check mode.
+            if vm_power_state == "running" and self.module.check_mode:
+                return True
+
+            # running state is required for guest shutdown.
+            if vm_power_state == "running":
+                if self.module.params['state_change_timeout'] == 0:
+                    self.xapi_session.xenapi.VM.clean_shutdown(self.vm_ref)
+                else:
+                    task_ref = self.xapi_session.xenapi.Async.VM.clean_shutdown(self.vm_ref)
+                    task_result = wait_for_task(self.module, task_ref, self.module.params['state_change_timeout'])
+
+                    if task_result:
+                        self.module.fail_json(msg="Guest shutdown task failed - '%s'!" % task_result)
+
+                self.vm_params['power_state'] = "halted"
+                changed_state = True
+            else:
+                self.module.fail_json(msg="VM must be in running state for guest shutdown! Current VM state is '%s'." % vm_power_state)
+        except XenAPI.Failure as f:
+            self.module.fail_json(msg="XAPI ERROR: %s" % f.details)
+
+        return changed_state
+
+    def rebootguest(self):
+        # Safety check.
+        if self.vm_ref is None:
+            self.module.fail_json(msg="Called rebootguest on non existing VM!")
+
+        changed_state = False
+
+        try:
+            # First check the current state of the VM.
+            vm_power_state = self.vm_params['power_state'].lower()
+
+            # Support for ansible check mode.
+            if vm_power_state == "running" and self.module.check_mode:
+                return True
+
+            # running state is required for guest reboot
```

Comment:
Could you please take a look at sub argument spec mentioned in https://github.com/ansible/ansible/pull/27119#issue-131585030 ? it would easy for user input validation and module simplification.

Reply:
Same for cdrom, custom_params and networks

Reply:
This looks great. I'll check into it. I wondered why such functionality was missing.

Reply:
Is this a requirement for merging the module to Ansible code base or just a suggestion for future development? This would require some rethinking of the code workflow and a lot of testing.

Reply:
Sub argument is not requirement but will make code concise and remove additional code used for user input.

Reply:
Definitely. This will be high on my TODO list but how soon I will be able to implement it and test remains to be seen.